lib/thrift/binary/framed/protocol_handler.ex

```diff
@@ -43,162 +51,257 @@ defmodule Thrift.Binary.Framed.ProtocolHandler do
 
-        {:ok, :reply, [message | serialized_reply]}
+        response_size = IO.iodata_length(serialized_message)
+        gauge(:response_size, response_size, state, method: method_name, result: "success")
+
+        send_reply([serialized_message | serialized_response], :continue, state)
 
       {:server_error, %TApplicationException{} = exc} ->
-        message = Protocol.Binary.serialize(:message_begin, {:exception, sequence_id, name})
-        serialized_exception = Protocol.Binary.serialize(:application_exception, exc)
+        finish_span(span, result: "error")
+
+        serialized_message =
+          Binary.serialize(:message_begin, {:exception, sequence_id, method_name})
 
-        {:error, {:server_error, [message | serialized_exception]}}
+        serialized_exception = Binary.serialize(:application_exception, exc)
 
-      :noreply ->
-        message = Protocol.Binary.serialize(:message_begin, {:reply, sequence_id, name})
+        response_size = IO.iodata_length(serialized_exception)
+        gauge(:response_size, response_size, state, method: method_name, result: "error")
 
-        {:ok, :reply, [message | <<0>>]}
+        send_reply([serialized_message | serialized_exception], :stop, state)
     end
   end
 
-  defp handle_thrift_message(
-         {:ok, {:oneway, _seq_id, name, args_binary}},
-         server_module,
-         handler_module
+  defp handle_message(
+         {:oneway, _sequence_id, method_name, args_binary},
+         %__MODULE__{server_module: server_module, handler_module: handler_module} = state
        ) do
-    spawn(server_module, :handle_thrift, [name, args_binary, handler_module])
-    {:ok, :reply, <<0>>}
+    spawn(fn ->
+      server_module.handle_thrift(method_name, args_binary, handler_module)
+    end)
+
+    send_reply(<<0>>, :continue, state)
+  end
+
+  defp send_reply(data, after_reply, %__MODULE__{transport: transport, socket: socket} = state) do
+    span = start_span(:send_reply, state)
+
+    case transport.send(socket, data) do
+      :ok when after_reply == :continue ->
+        finish_span(span, result: "success")
+        receive_message(state)
+
+      :ok when after_reply == :stop ->
+        finish_span(span, result: "success")
+        close(state)
+
+      {:error, closed} when closed in [:closed, :econnreset, :timeout] ->
+        finish_span(span, result: to_string(closed))
+        close(state)
+
+      {:error, error} ->
+        Logger.error(fn ->
+          format_log("send_reply", :ssl.format_error(error), state)
+        end)
+
+        finish_span(span, result: "error")
+        close(state)
+    end
+  end
+
+  defp close(%__MODULE__{transport: transport, socket: socket}) do
+    transport.close(socket)
+  end
+
+  defp start_span(metric_name, %__MODULE__{handler_module: handler_module}, tags \\ %{}) do
+    event_name = [Thrift, handler_module, metric_name]
+    time = System.monotonic_time()
+    metadata = Map.new(tags)
```

Comment:
Do we need to make a copy of `tags`?

Reply:
Tags is a keyword list when it is given as a parameter. Just for a little bit of readability really, to be able to write:
```
start_span(:call, state, method: method_name)
```
vs
```
start_span(:call, state, %{method: method_name})
```

Reply:
Ah, I missed that because of the default value for `tags` (which is `%{}`).

Reply:
> Ah, I missed that because of the default value for `tags` (which is `%{}`).

Yeah that's confusing, will change.